src/TextInputWidget.cc

```diff
@@ -170,6 +208,30 @@ FilteredTextEdit::textChanged()
         working_history_[history_index_] = toPlainText();
 }
 
+void
+FilteredTextEdit::receiveImage(const QPixmap &img, const QString &img_name)
+{
+        // Save image into temporary path to be loaded later.
+        QString img_path = QDir::tempPath() + '/' + img_name;
+        QSharedPointer<QFile> file{new QFile{img_path, this}};
+
+        if (!file->open(QIODevice::WriteOnly)) {
+                qDebug() << "Failed to open temporary image file:" << file->errorString();
+                previewDialog_->close();
+                return;
+        }
+        if (!img.save(file->fileName(), "PNG")) {
+                qDebug() << "Failed to save image data to:" << file->fileName();
+                previewDialog_->close();
+                return;
+        }
+        file->close();
+
+        emit image(file);
+
+        previewDialog_->close();
```

Comment:
Since you close the preview dialog no matter what, wouldn't it be simpler to do so in one place, in the dialog's implementation at the same time you emit `confirmImageUpload`? That way it's impossible to forget, and much less likely to be broken by future changes.